Multifactor authentication (WIP)

sasakalaba opened the pull request.

Commit:
Initial multifactor auth commit (WIP).

Commit:
Added MFA support on accounts.

 - Implemented first user story.

   (As a developer, I want to be able to add a phone number to an existing
   account and verify it, so I can support multifactor on existing accounts.)

 - Accounts can now add phone numbers and verify them via MFA.

Commit:
Added message customization.

 - Developers can now specifiy a custom SMS message.

Commit:
Added tests for the first user story.

 - added htmlcov dir to .gitignore (for using pytest coverage)
 - challenge_factor method on Factor no longer checks for '${code}'
   placeholder (invalid code will be reported from the api response)
 - challenge_factor method now returns a Challenge instance, not Factor
 - added live tests
 - added mock tests

Commit:
Updated MFA resources.

 - now supports google-authenticator
 - added additional mixins to new resources
 - added docstrings to all new resource classes
 - updated setting default message
 - added additional tests
 - replaced double quotes with single quotes in tests

avojnovicDk commented on stormpath/resources/challenge.py:
```diff
@@ -0,0 +1,46 @@
+)
+
+
+class Challenge(Resource, DeleteMixin, DictMixin, SaveMixin, StatusMixin):
```
`StatusMixin` nema smisla za ovaj resurs.

avojnovicDk commented on stormpath/resources/challenge.py:
```diff
@@ -0,0 +1,46 @@
+    """
+
+    writable_attrs = ('message', )
+
```
dodat `STATUS_` varijable (i `is_` funkcije?) - samo koje ce developeri koristit u kodu

avojnovicDk commented on stormpath/resources/challenge.py:
```diff
@@ -0,0 +1,46 @@
+            'factor': Factor,
+        }
+
+    def submit_challenge(self, code):
```
dovoljno je samo `submit` (s obzirom da smo vec na `Challenge` resursu)

avojnovicDk commented on stormpath/resources/challenge.py:
```diff
@@ -0,0 +1,46 @@
+        """
+        data = {'code': code}
+        self._store.update_resource(self.href, data)
+        self.refresh()
```
Cemu `refresh()` i zasto ne iskoristit `save()`?

avojnovicDk commented on stormpath/resources/factor.py:
```diff
@@ -0,0 +1,81 @@
+    """
+
+    writable_attrs = ('type', 'phone', 'challenge')
+    default_message = 'Your verification code is ${code}'
```
ovo je default na serveru, ne treba u SDK

avojnovicDk commented on stormpath/resources/factor.py:
```diff
@@ -0,0 +1,81 @@
+    Stormpath Factor resource.
+    """
+
+    writable_attrs = ('type', 'phone', 'challenge')
```
dodat `VERIFICATION_STATUS_` varijable i `is_verified()` funkcije

avojnovicDk replied:
dodat `TYPE_` varijable

avojnovicDk commented on stormpath/resources/factor.py:
```diff
@@ -0,0 +1,81 @@
+            'phone': Phone
+        }
+
+    def challenge_factor(self, message=default_message, code=None):
```
ne treba ova funkcija, dovoljan ce bit poziv `factor.challenges.create()`

avojnovicDk commented on stormpath/resources/factor.py:
```diff
@@ -0,0 +1,81 @@
+
+    def create(self, properties, challenge=True, expand=None, **params):
+        """
+        This methods will check for the challenge argument, set the proper
```
method, not methods

avojnovicDk commented on stormpath/resources/factor.py:
```diff
@@ -0,0 +1,81 @@
+        elif challenge and 'message' not in properties.get('challenge', {}):
+            # Set default message if one is not specified.
+            properties['challenge'] = {
+                'message': self.resource_class.default_message}
```
i validacija i setiranje defaulta ce su i tako na serveru

avojnovicDk commented on stormpath/resources/phone.py:
```diff
@@ -0,0 +1,31 @@
+    Stormpath Phone resource.
+    """
+
+    writable_attrs = ('number', )
```
i `verification_status` i `status` su writable

avojnovicDk replied:
- treba dodat varijable i funkcije za `verification_status`

avojnovicDk commented on stormpath/resources/factor.py:
```diff
@@ -0,0 +1,81 @@
+        :param bool challenge: Determines if a challenge is created on factor
+               creation.
+        """
+        if not challenge:
```
check / raise exception

Commit:
MFA refactor.

 - Factor:
    - added is_verified method on Factor
    - removed default message from Factor
    - invalid challenge parameter now raises a ValueError, instead of popping
      challenge from parameters
    - challenge_factor now calls challenges.create instead of
      _store.update_resource
    - added status, issuer and account_name to writable_attrs

 - Challenge:
    - replaced StatusMixin on Challenge with custom status methods
    - renamed submit_challenge() to submit() on Challenge
    - updated submit method on Challenge (now using save instead of
      _store.update_resource)

 - Phone:
    - Phone status and verification status are now editable

 - tests:
    - created MFA base test class
    - updated tests

avojnovicDk commented on tests/live/test_factor.py:
```diff
@@ -89,50 +81,59 @@ def test_create_without_challenge(self):
-        self.assertEqual(len(factor.challenges.items), 0)
-        challenge = factor.challenge_factor('${code}')
+        challenge = factor.challenge_factor(message='${code}')
+        factor.refresh()
```
Mozda ipak refreshat u `challenge_factor` funkciji? Not sure.
